### ISSUE TITLE
Not Writing Field_1 to the Database-Update

### ISSUE THREAD

**Issue by SDPHQ** [RagtagOpen/match-district#6]
Jill,

I figured out why this isn't writing to the database. Reading through the developer docs, I saw that connections are stored as arrays of record IDs (strings), even when there is only one connected record. Field 1 is stored as an array string because it's a connected field in Knack. 

Example:
Each of these is the ID of a record of the object to which this field connects.
// More than one connected record
["4fea069e9e8246001d000698", "4fea069e9e8246001d000699"]

// One connected record
["4fea069e9e8246001d000699"]

So, in order to be able to insert the information in field_1 into the database, the record ID of the connected field has to be queried and the put in. I.e. Send CA12 to Knack to figure out what the record ID is for that district, have that get sent back, and then that record id be placed into the field_1 instead of the text "CA12".

I found some examples of people playing around with similar issues here:
https://gist.github.com/ngalluzzo/8ca62a13fef3d2827901
https://support.knack.com/hc/en-us/community/posts/222366067-Javascript-Ajax-update-a-connected-field
https://support.knack.com/hc/en-us/community/posts/222366727-Can-t-retrieve-connected-record-field
https://support.knack.com/hc/en-us/community/posts/223976207-Update-connection-records

Apparently Knack has no way currently of just providing a list of records IDs. The connected records are found at object_4

Let me know if this doesn't make any sense and you want to hope on a phone call.

**Comment by jillh510** [comment]
In the record returned from the API call to Knack, I see several fields as "field_X_raw", which contains the data I sent via API, and "field_X", which seems to have had some processing done on the server. For instance:
record: Object
field_1: ""
field_1_raw: ["M", "A", "0", "7"] (4)
field_2: "<a href=\"mailto:kssksk@test.com\">kssksk@test.com</a>"
field_2_raw: {email: "kssksk@test.com"}

Is there some processing that occurs in the Knack view? (how does the page where you embed the Knack scene_21, view_33 handle it?)

It's also strange to me that Knack stores field_1_raw as an array of characters, where the other raw string fields are stored as strings. I send them all as strings in the http call.

**Comment by SDPHQ** [comment]
Yep, Knack is kind of strange that way. I think it’s because of the special nature of what’s contained in Field 1 because it’s a connector to another part of the database. Field 1 apparently is drawn from  object_4, field_56. That’s where the record ID is for the array string. 
_______________________________________________________________
Candis Mitchell
Co-Founder Sister District Project, Director of Information and Finance
candis.mitchell@sisterdistrict.com
www.sisterdistrict.com

**Comment by jillh510** [comment]
How does it work for the embedded Knack forms elsewhere on the site?

**Comment by SDPHQ** [comment]
Since the form is generated by Knack it autopopulates the form. If you go to our old volunteer form www.sisterdistrict.com/volunteer that first field that shows the district is the array field.

**Comment by jillh510** [comment]
Is this something that can be handled within the form using record rules?

https://support.knack.com/hc/en-us/articles/115002910307-Forms-Using-Record-Rules

**Comment by SDPHQ** [comment]
The record rules also don’t allow a text string to be written to an array string.  Its the first bullet under the 5.1 notes.

I can create a new field for the state + district number to go to and either export and import it in as a .csv which will write it correctly to the array or try to run a correction to the record after its been placed in the database.

Do you have particular language you’d like me to use to credit RagTag for the creation of this?
